web/signature_manager.js

```diff
@@ -614,6 +625,141 @@ class SignatureManager {
+      mustSmooth: false,
+      areContours,
+    });
+    if (!outlineData) {
+      return;
+    }
+
+    const { outline } = outlineData;
+    const svgFactory = new DOMSVGFactory();
+
+    const div = document.createElement("div");
+    const button = document.createElement("button");
+    button.addEventListener("click", () => {
+      this.#eventBus.dispatch("switchannotationeditorparams", {
+        source: this,
+        type: AnnotationEditorParamsType.CREATE,
+        value: {
+          signatureData: {
+            lines: {
+              curves,
+              thickness,
+              width,
+              height,
+            },
+            mustSmooth: false,
+            areContours,
+            description,
+            uuid,
+            heightInPage: 40,
+          },
+        },
+      });
+    });
+    div.append(button);
+    div.classList.add("toolbarAddSignatureButtonContainer");
+
+    const svg = svgFactory.create(1, 1, true);
+    button.append(svg);
+
+    const span = document.createElement("span");
+    button.append(span);
+
+    button.classList.add("toolbarAddSignatureButton");
+    button.type = "button";
+    button.title = span.textContent = description;
+    button.tabIndex = 0;
+
+    const path = svgFactory.createElement("path");
+    svg.append(path);
+    svg.setAttribute("viewBox", outline.viewBox);
+    svg.setAttribute("preserveAspectRatio", "xMidYMid meet");
+    if (areContours) {
+      svg.classList.add("contours");
+    }
+    path.setAttribute("d", outline.toSVGPath());
+
+    const deleteButton = document.createElement("button");
+    div.append(deleteButton);
+    deleteButton.classList.add("toolbarButton", "deleteButton");
+    deleteButton.setAttribute(
+      "data-l10n-id",
+      "pdfjs-editor-delete-signature-button"
+    );
+    deleteButton.title = "Delete";
+    deleteButton.type = "button";
+    deleteButton.tabIndex = 0;
+    deleteButton.addEventListener("click", async () => {
+      if (await this.#signatureStorage.delete(uuid)) {
+        div.remove();
+      }
+    });
+    const deleteSpan = document.createElement("span");
+    deleteButton.append(deleteSpan);
+    deleteSpan.setAttribute(
+      "data-l10n-id",
+      "pdfjs-editor-delete-signature-button-label"
+    );
+
+    this.#addSignatureToolbarButton.before(div);
+  }
+
+  async loadSignatures() {
+    if (
+      !this.#addSignatureToolbarButton ||
+      this.#addSignatureToolbarButton.previousElementSibling ||
+      !this.#signatureStorage
+    ) {
+      return;
+    }
+
+    if (!this.#loadSignaturesPromise) {
+      // The first call of loadSignatures() starts loading the signatures.
+      // The second one will wait until the signatures are loaded in the DOM.
+      this.#loadSignaturesPromise = this.#signatureStorage
+        .getAll()
+        .then(async signatures => [
+          signatures,
+          await Promise.all(
+            Object.values(signatures).map(({ signatureData }) =>
```

Comment:
Looking at this, I cannot help wonder if we could perhaps change the `SignatureStorage`-implementations to use `Map`s internally to avoid this sort of thing?